### ISSUE TITLE
CSRF Protection without starting session

### ISSUE THREAD

**Issue by Koc** [symfony/symfony#13464]
It would be nice add cookies csrf token storage. Use case: we are having form which available for anonymous users (and bots like googlebot). Session starts on each time when this page open. And too much files was generated with session data.

**Comment by jakzal** [comment]
This would kind of defeat the purpose of a csrf token, since the attacker would only need to make sure that both cookie and request tokens match.

**Comment by jaytaph** [comment]
I agree. There must be some stateful connection between the requests in order to set, and check the csrf tokens. 

I'm really curious how many anonymous requests are done? Google doesn't do many request on a single site, so quite possible you have issues in cleaning up old session-files? Other options might be to store sessions inside a database or persistent memory storage like redis?

**Comment by jakzal** [comment]
Giving it a bit more thought, it could be actually possible to implement by replacing session storage with some other kind of storage (like a database). 

Cookie could be used to carry user's identifier (like a session identifier, but without sessions heh). The identifier can then be used to fetch and refresh the token (stored in a db).

I think it's out of scope of the core framework though.

**Comment by merk** [comment]
Could you not use a HMAC generated with a secret only known by the server? You could also send a cookie with a client only secret to be included in the generated HMAC.. The server would then be able to verify the HMAC without starting the session.

**Comment by merk** [comment]
This does beg a question though - why do you need CSRF protection against XSS attacks for browser sessions which will not be authenticated?

**Comment by merk** [comment]
Sorry, never mind - stupid question. :)

**Comment by nicolas-grekas** [comment]
AFAIK, there is no need at all for any server side state. Comparing the cookie and the request token match is enough, provided they are truly random.
You could even have no cookie at all and use the `Origin` header in most situations.

**Comment by ircmaxell** [comment]
Cookies are fine for CSRF protection. The point is to prove that the request came from the user. In a cross-site context, an attacker cannot access cookies, nor can they set cookies. Therefore, if a cookie and a request token match, it had to come from a domain that could read and set both (assuming sufficient randomness in the token).

Therefore, that mode-of-operation for CSRF protection is perfectly acceptable. You don't need sessions to secure against CSRF.

**Comment by GromNaN** [comment]
We must specify the cookie format. I see 2 options:

**Option 1: One cookie**
All the CSRF tokens are URL encoded in an unique cookie.

```
csrf_tokens=s5513d32d09ca7%3Daek9-nde_j4ep7xAXL5e2YpBbDNCORoX4rQd8Qb6TTw%26s5513d3b0db59d%3DJyRxuq9sIYl3tZ1LkaLSylesRogqsAJN9R-j7uguMZE
```

This has the benefit of using only 1 cookie (easier for reverse proxy config), but in case of simultaneous requests, the cookie may be overwritten ; loosing the token of the 1st request.

**Option 2: Multiple cookies**
Create a distinct cookie for every CSRF token.

```
csrf_token_s5513d32d09ca7=aek9-nde_j4ep7xAXL5e2YpBbDNCORoX4rQd8Qb6TTw&csrf_token_s5513d3b0db59d=JyRxuq9sIYl3tZ1LkaLSylesRogqsAJN9R-j7uguMZE
```

This let each token live independently in its own cookie, avoiding most request collisions risks.

**Comment by uwej711** [comment]
From the OWASP wiki: Prevention Measures That Do NOT Work: using a Secret Cookie

see https://www.owasp.org/index.php/Cross-Site_Request_Forgery_%28CSRF%29_Prevention_Cheat_Sheet#Using_a_Secret_Cookie

**Comment by sjparkinson** [comment]
> From the OWASP wiki: Prevention Measures That Do NOT Work: using a Secret Cookie
> 
> see https://www.owasp.org/index.php/Cross-Site_Request_Forgery_%28CSRF%29_Prevention_Cheat_Sheet#Using_a_Secret_Cookie

If you have a read further down there's a section on the [double submit cookies](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_%28CSRF%29_Prevention_Cheat_Sheet#Double_Submit_Cookies) method that is perfectly valid.

**Comment by uwej711** [comment]
Right, just did that, so a cookie and a request parameter/hidden input are needed and the server checks they are the same. As I reread @ircmaxell comment now I realised he mentioned this but it wasn't so obvious when you just skim this issue ...

**Comment by gansbrest** [comment]
Curious to know what's the conclusion of this discussion, was the cookie support added? What's the recommended way to do it?

**Comment by htulipe** [comment]
Any updates on that issue ?

**Comment by nicolas-grekas** [comment]
@jderusse implemented cookie CSRF storage in #33171, but we decided against it because it appears to us that the complexity might not be worth the maintainance effort, especially when browsers implement or are implementing other countermeasures based on e.g. the `Origin` header. See the discussion in the PR for details.

**Comment by nicolas-grekas** [comment]
FTR, I'm experimenting on this topic in #58095
